<a href="https://colab.research.google.com/github/Thando367/Prediction-of-Product-Sales/blob/mainx/Project_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Prediction of Product Sales

<h6> Thandolwethu Siloyi

# Loading Data

In [711]:
# Mount google drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [712]:
# Import required packages
import pandas as pd
import numpy as np

# file path
df_filepath = "/content/drive/MyDrive/StudentFolders16wk20wk/CodingDojo/01-Fundamentals/Week02/Data/sales_predictions_2023.csv"

# reding sales_prediction 2023 data
df_sales = pd.read_csv(df_filepath)
df_sales

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0              FDA15        9.300          Low Fat         0.016047   
1              DRC01        5.920          Regular         0.019278   
2              FDN15       17.500          Low Fat         0.016760   
3              FDX07       19.200          Regular         0.000000   
4              NCD19        8.930          Low Fat         0.000000   
...              ...          ...              ...              ...   
8518           FDF22        6.865          Low Fat         0.056783   
8519           FDS36        8.380          Regular         0.046982   
8520           NCJ29       10.600          Low Fat         0.035186   
8521           FDN46        7.210          Regular         0.145221   
8522           DRG01       14.800          Low Fat         0.044878   

                  Item_Type  Item_MRP Outlet_Identifier  \
0                     Dairy  249.8092            OUT049   
1               Soft Drinks   48.2692            OUT018   
2                      Meat  141.6180            OUT049   
3     Fruits and Vegetables  182.0950            OUT010   
4                 Household   53.8614            OUT013   
...                     ...       ...               ...   
8518            Snack Foods  214.5218            OUT013   
8519           Baking Goods  108.1570            OUT045   
8520     Health and Hygiene   85.1224            OUT035   
8521            Snack Foods  103.1332            OUT018   
8522            Soft Drinks   75.4670            OUT046   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                          1999      Medium               Tier 1   
1                          2009      Medium               Tier 3   
2                          1999      Medium               Tier 1   
3                          1998         NaN               Tier 3   
4                          1987        High               Tier 3   
...                         ...         ...                  ...   
8518                       1987        High               Tier 3   
8519                       2002         NaN               Tier 2   
8520                       2004       Small               Tier 2   
8521                       2009      Medium               Tier 3   
8522                       1997       Small               Tier 1   

            Outlet_Type  Item_Outlet_Sales  
0     Supermarket Type1          3735.1380  
1     Supermarket Type2           443.4228  
2     Supermarket Type1          2097.2700  
3         Grocery Store           732.3800  
4     Supermarket Type1           994.7052  
...                 ...                ...  
8518  Supermarket Type1          2778.3834  
8519  Supermarket Type1           549.2850  
8520  Supermarket Type1          1193.1136  
8521  Supermarket Type2          1845.5976  
8522  Supermarket Type1           765.6700  

[8523 rows x 12 columns]

In [713]:
# displaying info for dataframe
df_sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


In [714]:
df_sales.head() # display first 5 row

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

# Data Cleaning

In [715]:
# checking number of rows and columns
df_sales.shape

(8523, 12)

There's 8523 rows and 12 columns

In [716]:
# checking the datatypes
df_sales.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

Data types are sorted according

# Checking Duplicated Data

In [717]:
# total number of duplicated data
tot_num_dupl = df_sales.duplicated().sum()
tot_num_dupl

0

No duplicates

# Checking Missing Values

In [718]:
# identify the missing values
df_sales.isnull().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

In [719]:
# identifying the percentage of missing data
df_sales.isna().sum() / len(df_sales)*100

Item_Identifier               0.000000
Item_Weight                  17.165317
Item_Fat_Content              0.000000
Item_Visibility               0.000000
Item_Type                     0.000000
Item_MRP                      0.000000
Outlet_Identifier             0.000000
Outlet_Establishment_Year     0.000000
Outlet_Size                  28.276428
Outlet_Location_Type          0.000000
Outlet_Type                   0.000000
Item_Outlet_Sales             0.000000
dtype: float64



1.   There are 1463 missing values in Iem_Weight column
2.   There are 2410 missing values in Outlet_Size column

In [720]:
# checking the missing values
df_sales.isnull().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

#Finding and Fixing Inconsistancies

In [721]:
# fix inconsistencies in categorical data.
# Check unique values in each categorical column
string_cols = df_sales.select_dtypes('object').columns

for col in string_cols:
  print(f'Unique values for {col}')
  print(df_sales[col].value_counts())
  print('\n')


Unique values for Item_Identifier
FDW13    10
FDG33    10
NCY18     9
FDD38     9
DRE49     9
         ..
FDY43     1
FDQ60     1
FDO33     1
DRF48     1
FDC23     1
Name: Item_Identifier, Length: 1559, dtype: int64


Unique values for Item_Fat_Content
Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64


Unique values for Item_Type
Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64


Unique values for Outlet_Identifier
OUT027    935
OUT013    932
OUT049    930
OUT046    

In [722]:
# checking inconsistancies for numerical
num_cols = df_sales.select_dtypes('number').columns

for number in num_cols:
  print(f'Unique values for{number}')
  print(df_sales[number].value_counts())
  print('\n')

Unique values forItem_Weight
12.150    86
17.600    82
13.650    77
11.800    76
15.100    68
          ..
7.275      2
7.685      1
9.420      1
6.520      1
5.400      1
Name: Item_Weight, Length: 415, dtype: int64


Unique values forItem_Visibility
0.000000    526
0.076975      3
0.162462      2
0.076841      2
0.073562      2
           ... 
0.013957      1
0.110460      1
0.124646      1
0.054142      1
0.044878      1
Name: Item_Visibility, Length: 7880, dtype: int64


Unique values forItem_MRP
172.0422    7
170.5422    6
196.5084    6
188.1872    6
142.0154    6
           ..
97.3384     1
83.1934     1
96.6752     1
152.6682    1
75.4670     1
Name: Item_MRP, Length: 5938, dtype: int64


Unique values forOutlet_Establishment_Year
1985    1463
1987     932
1999     930
1997     930
2004     930
2002     929
2009     928
2007     926
1998     555
Name: Outlet_Establishment_Year, dtype: int64


Unique values forItem_Outlet_Sales
958.7520     17
1342.2528    16
703.0848     15
1845

There are inconsistancies in Item_Fat_Column


*   LF must be changed to Low fat
*   low fat to Low Fat
*   reg to Regular

In [723]:
# fixing inconsistancies
# for Item_Fat_Content
def fat_content(value):
    value = value.lower()
    if 'low fat' in value:
        return 'Low Fat'
    elif 'reg' in value:
        return 'Regular'
    elif 'lf' in value:
        return 'Low Fat'
    else:
        return value

df_sales['Item_Fat_Content'] = df_sales['Item_Fat_Content'].apply(fat_content)


In [724]:
# fixing inconsistancies for Outlet_Type
def outlet_type(value):
  value = value.lower()
  if 'supermarket type1' in value:
    return 'Supermarket Type1'
  elif 'supermarket type2' in value:
    return 'Supermarket Type2'
  elif 'supermarket type3' in value:
    return 'Supermarket Type3'
  elif 'grocery store' in value:
    return 'Grocery Store'
  else:
        return value

# applying function
df_sales['Outlet_Type'] = df_sales['Outlet_Type'].apply(outlet_type)
df_sales

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0              FDA15        9.300          Low Fat         0.016047   
1              DRC01        5.920          Regular         0.019278   
2              FDN15       17.500          Low Fat         0.016760   
3              FDX07       19.200          Regular         0.000000   
4              NCD19        8.930          Low Fat         0.000000   
...              ...          ...              ...              ...   
8518           FDF22        6.865          Low Fat         0.056783   
8519           FDS36        8.380          Regular         0.046982   
8520           NCJ29       10.600          Low Fat         0.035186   
8521           FDN46        7.210          Regular         0.145221   
8522           DRG01       14.800          Low Fat         0.044878   

                  Item_Type  Item_MRP Outlet_Identifier  \
0                     Dairy  249.8092            OUT049   
1               Soft Drinks   48.2692            OUT018   
2                      Meat  141.6180            OUT049   
3     Fruits and Vegetables  182.0950            OUT010   
4                 Household   53.8614            OUT013   
...                     ...       ...               ...   
8518            Snack Foods  214.5218            OUT013   
8519           Baking Goods  108.1570            OUT045   
8520     Health and Hygiene   85.1224            OUT035   
8521            Snack Foods  103.1332            OUT018   
8522            Soft Drinks   75.4670            OUT046   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                          1999      Medium               Tier 1   
1                          2009      Medium               Tier 3   
2                          1999      Medium               Tier 1   
3                          1998         NaN               Tier 3   
4                          1987        High               Tier 3   
...                         ...         ...                  ...   
8518                       1987        High               Tier 3   
8519                       2002         NaN               Tier 2   
8520                       2004       Small               Tier 2   
8521                       2009      Medium               Tier 3   
8522                       1997       Small               Tier 1   

            Outlet_Type  Item_Outlet_Sales  
0     Supermarket Type1          3735.1380  
1     Supermarket Type2           443.4228  
2     Supermarket Type1          2097.2700  
3         Grocery Store           732.3800  
4     Supermarket Type1           994.7052  
...                 ...                ...  
8518  Supermarket Type1          2778.3834  
8519  Supermarket Type1           549.2850  
8520  Supermarket Type1          1193.1136  
8521  Supermarket Type2          1845.5976  
8522  Supermarket Type1           765.6700  

[8523 rows x 12 columns]

In [725]:
#checking the number of numerical columns
numerical_columns = df_sales.select_dtypes(include='number').columns
numerical_columns

Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year', 'Item_Outlet_Sales'],
      dtype='object')

In [726]:
df_sales.describe()

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year  \
count  7060.000000      8523.000000  8523.000000                8523.000000   
mean     12.857645         0.066132   140.992782                1997.831867   
std       4.643456         0.051598    62.275067                   8.371760   
min       4.555000         0.000000    31.290000                1985.000000   
25%       8.773750         0.026989    93.826500                1987.000000   
50%      12.600000         0.053931   143.012800                1999.000000   
75%      16.850000         0.094585   185.643700                2004.000000   
max      21.350000         0.328391   266.888400                2009.000000   

       Item_Outlet_Sales  
count        8523.000000  
mean         2181.288914  
std          1706.499616  
min            33.290000  
25%           834.247400  
50%          1794.331000  
75%          3101.296400  
max         13086.964800

In [727]:
# Drop Unwanted columns
df_sales = df_sales.drop(columns = ['Outlet_Identifier', 'Item_Identifier', 'Outlet_Establishment_Year'])
# Display the .head to confirm the change
df_sales.head()

Item_Weight Item_Fat_Content  Item_Visibility              Item_Type  \
0         9.30          Low Fat         0.016047                  Dairy   
1         5.92          Regular         0.019278            Soft Drinks   
2        17.50          Low Fat         0.016760                   Meat   
3        19.20          Regular         0.000000  Fruits and Vegetables   
4         8.93          Low Fat         0.000000              Household   

   Item_MRP Outlet_Size Outlet_Location_Type        Outlet_Type  \
0  249.8092      Medium               Tier 1  Supermarket Type1   
1   48.2692      Medium               Tier 3  Supermarket Type2   
2  141.6180      Medium               Tier 1  Supermarket Type1   
3  182.0950         NaN               Tier 3      Grocery Store   
4   53.8614        High               Tier 3  Supermarket Type1   

   Item_Outlet_Sales  
0          3735.1380  
1           443.4228  
2          2097.2700  
3           732.3800  
4           994.7052

In [728]:
# new imports
# Import packages
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns
pd.set_option('display.max_columns',100)
import missingno
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.pipeline import make_pipeline
from sklearn.impute import SimpleImputer
from sklearn.compose import make_column_selector
import pandas as pd
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer

# Set pandas as the default output for sklearn
from sklearn import set_config
set_config(display='pandas')


In [729]:
# define feature X and exclude name column
target = 'Item_Outlet_Sales'
X = df_sales.drop(columns=[target]).copy()

# define target y as Item_Outlet_Sales
y = df_sales[target].copy()

In [730]:
# performing the train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)


In [731]:
X.head() # display X columns

Item_Weight Item_Fat_Content  Item_Visibility              Item_Type  \
0         9.30          Low Fat         0.016047                  Dairy   
1         5.92          Regular         0.019278            Soft Drinks   
2        17.50          Low Fat         0.016760                   Meat   
3        19.20          Regular         0.000000  Fruits and Vegetables   
4         8.93          Low Fat         0.000000              Household   

   Item_MRP Outlet_Size Outlet_Location_Type        Outlet_Type  
0  249.8092      Medium               Tier 1  Supermarket Type1  
1   48.2692      Medium               Tier 3  Supermarket Type2  
2  141.6180      Medium               Tier 1  Supermarket Type1  
3  182.0950         NaN               Tier 3      Grocery Store  
4   53.8614        High               Tier 3  Supermarket Type1

<h1>Create a preprocessing object to prepare the dataset for Machine Learning

*   Numeric
*   Categorical



<h3> 1. Numerical

In [732]:
# creating a numerical data selector
num_cols = X_train.select_dtypes('number').columns
num_cols


Index(['Item_Weight', 'Item_Visibility', 'Item_MRP'], dtype='object')

In [733]:
#instantiate preprocessors
miss_imputer = SimpleImputer(strategy='median')
scaler = StandardScaler()

#creating pipeline
num_pipeline = make_pipeline(miss_imputer, scaler)


# Making a numeric tuple for ColumnTransformer
num_tuple = ('numeric', num_pipeline, num_cols)
num_tuple

('numeric',
 Pipeline(steps=[('simpleimputer', SimpleImputer(strategy='median')),
                 ('standardscaler', StandardScaler())]),
 Index(['Item_Weight', 'Item_Visibility', 'Item_MRP'], dtype='object'))

<h4> 2. Categoric

In [734]:
#from pandas.core.dtypes.dtypes import Ordered

# Categorical - column names
cat_cols = X_train.select_dtypes('object').columns
cat_cols

Index(['Item_Fat_Content', 'Item_Type', 'Outlet_Size', 'Outlet_Location_Type',
       'Outlet_Type'],
      dtype='object')

In [735]:
# Instantiate the preprocessors
imputer = SimpleImputer(strategy='constant', fill_value='NA')
one_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')

# create a pipeline with imputer and onehotencoder
cat_pipeline = make_pipeline(imputer, one_encoder)

# Making a ohe_tuple for ColumnTransformer
cat_tuple = ('categorical', cat_pipeline, cat_cols)
cat_tuple

('categorical',
 Pipeline(steps=[('simpleimputer',
                  SimpleImputer(fill_value='NA', strategy='constant')),
                 ('onehotencoder',
                  OneHotEncoder(handle_unknown='ignore', sparse_output=False))]),
 Index(['Item_Fat_Content', 'Item_Type', 'Outlet_Size', 'Outlet_Location_Type',
        'Outlet_Type'],
       dtype='object'))

In [736]:
# making a column transformer
preprocessor = ColumnTransformer([num_tuple, cat_tuple],
                                 verbose_feature_names_out=False)
preprocessor

ColumnTransformer(transformers=[('numeric',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(strategy='median')),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 Index(['Item_Weight', 'Item_Visibility', 'Item_MRP'], dtype='object')),
                                ('categorical',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(fill_value='NA',
                                                                strategy='constant')),
                                                 ('onehotencoder',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False))]),
                                

In [737]:
# Fit the preprocessor on X_train data
preprocessor.fit(X_train)

ColumnTransformer(transformers=[('numeric',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(strategy='median')),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 Index(['Item_Weight', 'Item_Visibility', 'Item_MRP'], dtype='object')),
                                ('categorical',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(fill_value='NA',
                                                                strategy='constant')),
                                                 ('onehotencoder',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False))]),
                                

In [738]:
# Transforming the training and test data
X_train_tf = preprocessor.transform(X_train)
X_test_tf = preprocessor.transform(X_test)


In [739]:
import numpy as np

print("Shape:", X_train_tf.shape)
print("Data Type:", X_train_tf.dtype)


Shape: (6392, 32)
Data Type: float64


In [740]:
# function
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
def regression_metrics(reg,X_test,y_test, label='Test Data'):

  # Get predictions
  y_pred = reg.predict(X_test)

  # Get metrics
  mae = mean_absolute_error(y_test, y_pred)
  mse = mean_squared_error(y_test, y_pred)
  rmse = mean_squared_error(y_test, y_pred, squared=False)
  r_squared = r2_score(y_test, y_pred)

  # Print Result with label
  header = "---"*20
  print(header,f"Regresion Metrics: {label}",header, sep='\n')
  print(f"MAE = {mae:,.2f}")
  print(f"MSE = {mse:,.2f}")
  print(f"RMSE = {rmse:,.2f}")
  print(f"R^2 = {r_squared:,.2f}")



def evaluate_regression(reg, X_train, y_train, X_test, y_test):

  # Get results for training data
  regression_metrics(reg, X_train, y_train, label='Training Data' )

  # Get results for test data
  regression_metrics(reg, X_test, y_test, label='Test Data' )

# Modeling

### Linear Modeling

In [741]:
from sklearn.linear_model import LinearRegression

In [742]:
# creating linear regression
linear_reg = LinearRegression()
linear_reg

LinearRegression()

In [743]:
# creating pipeline
linear_reg_pipe = make_pipeline(preprocessor, linear_reg)
# fitting the model
linear_reg_pipe.fit(X_train, y_train);
# evaluation
evaluate_regression(linear_reg_pipe, X_train, y_train, X_test, y_test)

------------------------------------------------------------
Regresion Metrics: Training Data
------------------------------------------------------------
MAE = 847.02
MSE = 1,298,680.69
RMSE = 1,139.60
R^2 = 0.56
------------------------------------------------------------
Regresion Metrics: Test Data
------------------------------------------------------------
MAE = 805.64
MSE = 1,197,464.73
RMSE = 1,094.29
R^2 = 0.57




*   R^2 (R-Squared) :
        *   Training data - 0,56
        *   Test data - 0,57

*   The model is not Overfitting at the same time it is not Underfitting. The values for both training and test are close. Provides a balanced fit to the data.



## Random Forest

In [744]:
from sklearn.ensemble import RandomForestRegressor

In [746]:
# creating Random Forest Model
random_forest = RandomForestRegressor(random_state=42)
# creating pipeline
rf_pipeline = make_pipeline(preprocessor, random_forest)
# fitting the model
rf_pipeline.fit(X_train, y_train);
# evaluate
evaluate_regression(rf_pipeline, X_train, y_train, X_test, y_test)


------------------------------------------------------------
Regresion Metrics: Training Data
------------------------------------------------------------
MAE = 299.94
MSE = 186,779.10
RMSE = 432.18
R^2 = 0.94
------------------------------------------------------------
Regresion Metrics: Test Data
------------------------------------------------------------
MAE = 774.10
MSE = 1,252,845.36
RMSE = 1,119.31
R^2 = 0.55




*   R^2 (R-Squared) :
        *   Training data - 0.94
        *   Test data - 0.55

*   The training R-Squared value is greater than the tes R-Squared value. The is a huge gap between the training and test R-squared values, Random Forest might be overfitting the training data.

*   To improve the model's performance we need to do fine-tuning.





## GridSearchCV (Tuning)

In [747]:
from sklearn.model_selection import train_test_split, GridSearchCV

In [748]:
# Options for tuning
rf_pipeline.get_params()

{'memory': None,
 'steps': [('columntransformer',
   ColumnTransformer(transformers=[('numeric',
                                    Pipeline(steps=[('simpleimputer',
                                                     SimpleImputer(strategy='median')),
                                                    ('standardscaler',
                                                     StandardScaler())]),
                                    Index(['Item_Weight', 'Item_Visibility', 'Item_MRP'], dtype='object')),
                                   ('categorical',
                                    Pipeline(steps=[('simpleimputer',
                                                     SimpleImputer(fill_value='NA',
                                                                   strategy='constant')),
                                                    ('onehotencoder',
                                                     OneHotEncoder(handle_unknown='ignore',
                                   

In [749]:
# Define param grid with options to try
params = {'randomforestregressor__max_depth': [None,10,15,20],
          'randomforestregressor__n_estimators':[10,100,150,200],
          'randomforestregressor__min_samples_leaf':[2,3,4],
          'randomforestregressor__max_features':['sqrt','log2',None],
          'randomforestregressor__oob_score':[True,False],
          }

In [754]:
# instantiate the GridSearch
grid_s = GridSearchCV(rf_pipeline, params, n_jobs=-1, cv = 3, verbose = 1)

# Fitting the GridSearch on the training data
grid_s.fit(X_train, y_train)

Fitting 3 folds for each of 288 candidates, totalling 864 fits


GridSearchCV(cv=3,
             estimator=Pipeline(steps=[('columntransformer',
                                        ColumnTransformer(transformers=[('numeric',
                                                                         Pipeline(steps=[('simpleimputer',
                                                                                          SimpleImputer(strategy='median')),
                                                                                         ('standardscaler',
                                                                                          StandardScaler())]),
                                                                         Index(['Item_Weight', 'Item_Visibility', 'Item_MRP'], dtype='object')),
                                                                        ('categorical',
                                                                         Pipeline(steps=[('simpleimputer',
                                             

In [755]:
# getting the best params
grid_s.best_params_

{'randomforestregressor__max_depth': 10,
 'randomforestregressor__max_features': None,
 'randomforestregressor__min_samples_leaf': 3,
 'randomforestregressor__n_estimators': 100,
 'randomforestregressor__oob_score': True}

In [759]:
# Defining best model
rf_best = grid_s.best_estimator_


In [760]:
# refit
evaluate_regression(rf_best, X_train, y_train, X_test, y_test)

------------------------------------------------------------
Regresion Metrics: Training Data
------------------------------------------------------------
MAE = 656.21
MSE = 875,839.04
RMSE = 935.86
R^2 = 0.70
------------------------------------------------------------
Regresion Metrics: Test Data
------------------------------------------------------------
MAE = 736.06
MSE = 1,121,801.30
RMSE = 1,059.15
R^2 = 0.59



Tuned Random Forest:

*   R^2 (R-Squared) :
        *   Training data - 0.70
        *   Test data - 0.59

*   Compared to the default model Tuned Random Forest have a better performance.

*   The gap between trsining and test R-squared values has been reduced.






## CRISP-DM Phase 5 - Evaluation

<h3> Model Recommendation

**Interpretation of R-Squared:**
R-Squared is a measure that enables us to see how our model's forecasts correspond to the outcomes in practice. Consider this way to measure the model's capacity for explaining variation data.

1. Random Forest:
R-squared (R2) on Test Data: 0.55
MAE on Test Data: 774.10
Reason The model explains about 55% of the variation in sales using the patterns it learned from the data. On average, its predictions are off by around 774.10 compared to the actual sales values.


2. GridSearchCV-tuned Random Forest:
R-squared (R^2) on Test Data: 0.55
MAE on Test Data: 774.10
Reason: This tuned model improved its R-squared to about 59%, suggesting it's explaining more of the sales variation. The average prediction error is now around 736.06, which is a bit better


3. Linear Regression: R-squared (R^2) on Test Data: 0.57
MAE on Test Data: 805.64
Reason: This model explains about 57% of the variation in sales. It's predictions are, on average, about 805.65 away from the real sales values

Selection of MAE Metric:
We're focusing on Mean Absolute Error (MAE) because it gives us a clear picture of how close our model's predictions are to the actual sales amounts. Lower MAE is better, as it indicates smaller prediction errors in terms of dollars. Stakeholders can easily relate to this concept.

Training vs. Test Scores and Model Fit:
Comparing the models, we see that the Random Forest model had a larger gap between training (0.94) and test (0.55) R-squared values, indicating potential overfitting. The GridSearchCV-tuned Random Forest reduced this gap (0.70 training, 0.59 test), suggesting improved generalization. The Linear Regression model is in between (0.56 training, 0.57 test), indicating a balanced fit.

In summary, the GridSearchCV-tuned Random Forest model is the best choice among the three. It explains more of the sales patterns and its predictions are, on average, around 736.06 away from the actual sales values. The model's performance on new data is consistent with its performance on the training data, indicating balanced fitting and better generalization.